services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/taydennys/TaydennysAttachmentService.kt

```diff
@@ -0,0 +1,98 @@
+
+private val logger = KotlinLogging.logger {}
+
+@Service
+class TaydennysAttachmentService(
+    private val metadataService: TaydennysAttachmentMetadataService,
+    private val taydennysRepository: TaydennysRepository,
+    private val attachmentContentService: TaydennysAttachmentContentService,
+    private val scanClient: FileScanClient,
+) {
+    fun addAttachment(
+        taydennysId: UUID,
+        attachmentType: ApplicationAttachmentType,
+        attachment: MultipartFile,
+    ): TaydennysAttachmentMetadataDto {
+        logger.info {
+            "Adding attachment to täydennys, taydennysId = $taydennysId, " +
+                "attachment name = ${attachment.originalFilename}, size = ${attachment.bytes.size}, " +
+                "content type = ${attachment.contentType}"
+        }
+        AttachmentValidator.validateSize(attachment.bytes.size)
+        val filename = AttachmentValidator.validFilename(attachment.originalFilename)
+        AttachmentValidator.validateExtensionForType(filename, attachmentType)
+        val taydennys = findTaydennys(taydennysId)
+
+        val contentType = AttachmentValidator.ensureMediaType(attachment.contentType)
+        scanAttachment(filename, attachment.bytes)
+        metadataService.ensureRoomForAttachment(taydennys)
+
+        val newAttachment =
+            saveAttachment(taydennys, attachment.bytes, filename, contentType, attachmentType)
+
+        return newAttachment.toDto()
+    }
+
+    fun saveAttachment(
+        taydennys: TaydennysIdentifier,
+        content: ByteArray,
+        filename: String,
+        contentType: MediaType,
+        attachmentType: ApplicationAttachmentType,
+    ): TaydennysAttachmentMetadata {
+        logger.info { "Saving attachment content for täydennys. ${taydennys.logString()}" }
+        val blobPath = attachmentContentService.upload(filename, contentType, content, taydennys.id)
+        logger.info { "Saving attachment metadata for täydennys. ${taydennys.logString()}" }
+        val newAttachment =
+            try {
+                metadataService.create(
+                    filename,
+                    contentType.toString(),
+                    content.size.toLong(),
+                    blobPath,
+                    attachmentType,
+                    taydennys.id,
+                )
+            } catch (e: Exception) {
+                logger.error(e) {
+                    "Attachment metadata save failed, deleting attachment content $blobPath"
+                }
+                attachmentContentService.delete(blobPath)
+                throw e
+            }
+        logger.info {
+            "Added attachment metadata ${newAttachment.id} and content $blobPath for täydennys. ${taydennys.logString()}"
+        }
+        return newAttachment
+    }
+
+    private fun findTaydennys(taydennysId: UUID): TaydennysIdentifier =
+        taydennysRepository.findByIdOrNull(taydennysId)?.toMetadata()
```

Comment:
`toMetaData()` isn't called under a transaction, but it uses `taydennyspyynto`, which is a lazily loaded field. Is there a risk this might fail at some point?

Reply:
I see. It's clearly in the [javadocs](https://jakarta.ee/specifications/persistence/3.2/apidocs/jakarta.persistence/jakarta/persistence/fetchtype). The persistence provider i.e. Hibarnate has to obey when the fetch type is `EAGER`, but `LAZY` is just a hint. So it can do eager fetches even when the fetch type is `LAZY`. When I turned SQL logging on, it clearly did an eager fetch here.

Since we're excepting the fetch to be eager here, maybe it would be better to set the `taydennyspyynto` relationship to be explicitly `EAGER`?

Reply:
Can't we just return the `TaydennysEntity`, since it already implements the `TaydennysIdentifier` interface?

Reply:
Interestingly, if I change `TaydennysEntity.taydennyspyynto` to be `EAGER`. it breaks integration tests on the other relationship `yhteystiedot`: 
```
org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: fi.hel.haitaton.hanke.taydennys.TaydennysyhteystietoEntity.yhteyshenkilot: could not initialize proxy - no Session

	at org.hibernate.collection.spi.AbstractPersistentCollection.throwLazyInitializationException(AbstractPersistentCollection.java:636)
	at org.hibernate.collection.spi.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:219)
	at org.hibernate.collection.spi.AbstractPersistentCollection.readSize(AbstractPersistentCollection.java:150)
	at org.hibernate.collection.spi.PersistentBag.size(PersistentBag.java:353)
	at kotlin.collections.CollectionsKt__IterablesKt.collectionSizeOrDefault(Iterables.kt:39)
	at fi.hel.haitaton.hanke.hakemus.YhteystietoEntity$DefaultImpls.toDomain(YhteystietoEntity.kt:47)
	at fi.hel.haitaton.hanke.taydennys.TaydennysyhteystietoEntity.toDomain(TaydennysyhteystietoEntity.kt:22)
	at fi.hel.haitaton.hanke.taydennys.TaydennysEntity.toDomain(TaydennysEntity.kt:52)
	at fi.hel.haitaton.hanke.factory.TaydennysFactory.updateTaydennys(TaydennysFactory.kt:110)
	at fi.hel.haitaton.hanke.factory.TaydennysFactory.updateJohtoselvitysTaydennys(TaydennysFactory.kt:118)
	at fi.hel.haitaton.hanke.taydennys.TaydennysServiceITest$SendTaydennys.sends the data to Allu with list of changed fields(TaydennysServiceITest.kt:386)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

Reply:
We split the attachment services into several services to avoid having the database transaction active while we're uploading the attachment file to the antivirus service and to Blob Storage.

Now you're undoing that work for the täydennys attachment because there's a failing test data factory method. Maybe you could try to fix the factory instead?

Reply:
Looks like the only required changes are adding `@Transactional` to `TaydennysFactory.updateTaydennys()` and `TaydennysFactory.updateJohtoselvitysTaydennys()`.

This breaks the `save()` method. It's probably because it uses a dependency in the parameter default value. The `taydennyspyynto` parameter is not used anywhere so it can be moved to a value in the method body.

Reply:
Thank you for taking time to advice me in this.

I did the modifications you mentioned except that `TaydennysFactory.updateTaydennys()` does not need to have `@Transactional` because it is only called from `updateJohtoselvitysTaydennys()`.

Reply:
And `TaydennysEntity.taydennyspyynto` is now `EAGER`.

Reply:
`TaydennysFactory.updateTaydennys()` either needs to be private or have `@Transactional` as well.

Reply:
Made it private.